Bottom: 1ea020c9976342d490caf5fca543a8f59f1005d6
Top:    fbbbdb29c84ef5e9d710ac88d417bca1bfe06928
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-07-25 14:09:09 +0100

Draw pin / pad names first, so they don't get masked by the pad

Since we use stencil masking to ensure we don't redraw an area
multiple times, confusingly, we need to draw the text first to
ensure it isn't occluded by the pad its-self.

HACK:

We probably draw the names multiple times - as we still use the
common draw.c code which assumes it can draw labels on top of
the pads.


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 7fa972a..5ad0729 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -1149,7 +1149,21 @@ draw_pin (PinTypePtr pin, bool draw_hole)
 static int
 pin_callback (const BoxType * b, void *cl)
 {
-  draw_pin ((PinType *)b, false);
+  PinTypePtr pin = (PinTypePtr) b;
+
+  if (!TEST_FLAG (HOLEFLAG, pin) && TEST_FLAG (DISPLAYNAMEFLAG, pin))
+    _draw_pv_name (pin);
+  draw_pin (pin, false);
+  return 1;
+}
+
+static int
+pin_name_callback (const BoxType * b, void *cl)
+{
+  PinTypePtr pin = (PinTypePtr) b;
+
+  if (!TEST_FLAG (HOLEFLAG, pin) && TEST_FLAG (DISPLAYNAMEFLAG, pin))
+    _draw_pv_name (pin);
   return 1;
 }
 
@@ -1266,8 +1280,11 @@ pad_callback (const BoxType * b, void *cl)
   PadTypePtr pad = (PadTypePtr) b;
   int *side = cl;
 
-  if (ON_SIDE (pad, *side))
+  if (ON_SIDE (pad, *side)) {
+    if (TEST_FLAG (DISPLAYNAMEFLAG, pad))
+      draw_pad_name (pad);
     draw_pad (pad);
+  }
   return 1;
 }
 
@@ -1601,6 +1618,9 @@ DrawLayerGroup (int group, const BoxType * screen)
 
       /* Draw pins, vias and pads on this layer */
       if (!global_view_2d && rv) {
+        if (PCB->PinOn &&
+            (group == solder_group || group == component_group))
+          r_search (PCB->Data->pin_tree, screen, NULL, pin_name_callback, Layer);
         if (PCB->PinOn) r_search (PCB->Data->pin_tree, screen, NULL, pin_inlayer_callback, Layer);
         if (PCB->ViaOn) r_search (PCB->Data->via_tree, screen, NULL, via_inlayer_callback, Layer);
         if (PCB->PinOn && group == component_group)
